README.md
# VC65X
6502, 65C02 and 65C816 Development Environment for Visual Studio Code

![re](images/screenshot1.png)
## Features

* Syntax highlighting for ACME assembler files
* Integration of the ACME assembler
* Full debugging support for MCS6502 CPU

## Requirements

There are no additional requirements or dependencies to operate this extension.

## Introduction and Basic Usage

The VC65X extension provides a convienient editing, build and run environment. This is done by providing syntax highlighting, automatic background compilation using the ACME assembler, an embedded 6502 CPU emulator.

### Syntax Highlighting

Support for ACME assember syntax is provided.

### Background Compilation

Whenever a `.asm` source file is modified and saved, it is automatically compiled to a plain `.bin` program file.

If compilation is successful, that program file can either be run/debugged with the embedded debugger based on a defined launch configuration (see *Debugger Launch Configuration*).

If compilation fails, the ACME outputs are shown in the diagnostics view.

### MCS6502 CPU Emulator

The VC65X extension comes with a built-in 6502 CPU emulator that allows very fast edit-build-run cycles. It purely executes 6502 machine code as fast as possible - and integrates nicely to the Visual Studio Code debugger interface.

An active 65XX debugging session allows you to:

- define breakpoints
- inspect registers, addresses, values
- get hover information for many elements of the source code

## Preferences/Settings

To setup the VC65X development environment, go to Preferences>Settings and check the following settings:

> VC65X: Assembler Path

Path to assembler executable.

Example: `C:\Tools\vc65x\acme\acme.exe`

> VC65X: Assembler Arguments

Additional assembler command line options.

> VC65X: Auto Build

Enable auto build before running or debugging.

> VC65X: Background Build

Enable background build after source changes.

## Debugger Launch Configuration

In order to run a compiled 65XX program (`.bin`) using the embedded 6502 CPU emulator, you have to create a launch configuration. Here is a small example:

    {
        "version": "0.2.0",
        "configurations": [
            {
                "type": "asm",
                "request": "launch",
                "name": "Launch Program",
                "pc": "0x0100",
                "base": "0x0000",
                "arch":"MCS6502",
                "binary": "C:\\Work\\vc65x\\demo1\\.cache\\src\\test.bin"
            }
        ]
    }

> `type`: Launch type

Always needs to be "asm".

> `request`: Request type

Always use "launch" here.

> `name`: Launch configuration name

Any name you want to use is fine.

> `binary`: Path to a compiled 65XX program

The default build output path is ".cache" within the workspace root folder.

> `pc`: Optional parameter to overwrite the start address of the 65X program

A 16-bit address in decimal or $hexadecimal form.

> `base`: Optional parameter to overwrite the base load address of the 65X program

A 16-bit address in decimal or $hexadecimal form.

> `arch`: Optional parameter to overwrite the CPU architecture of the emulator.

Select from "MCS6502", "R65C02", "W65C02S" or "W65C816S".

## Open Source

This package includes open source from other developers and I would like to thank all of those:

* Gregory Estrade - 6502.js: It was great to have your 6502 emulator to form the core of the debugger. Thank you for compressing the 6502 cpu in such a nice piece of software!
* Tony Landi - Acme Cross Assembler (C64): I started with the basic syntax definition for ACME from your package. Thanks for starting that!

## ToDo
- [X] Implement 65C02 variant
- [ ] Implement 65816 variant
- [ ] add [v6502r](https://github.com/floooh/v6502r) as emulation option
## Ideas Taken From
* Captain JiNX - VSCode KickAss (C64)
* Janne Hellsten - c64jasm

## Links

* VS64 - The C64 Development Environment : https://github.com/rosc77/vs64
* The ACME Cross-Assembler: https://sourceforge.net/projects/acme-crossass/
* VICE, the Versatile Commodore Emulator: http://vice-emu.sourceforge.net/
* C64 65XE Debugger: https://sourceforge.net/projects/c64-debugger/
* Cycle-accurate 6502 emulator in Javascript: https://github.com/Torlus/6502.js
* set of functional tests for the 6502/65C02 type processors: https://github.com/Klaus2m5/6502_65C02_functional_tests
* ideas for system-level support : https://github.com/LowLevelJavaScript/16-Bit-Virtual-Machine
* nice extension documentation ideas: https://github.com/prb28/vscode-amiga-assembly


CHANGELOG.md
# Change Log
All notable changes to the **VS65 Development Environment** extension will be documented in this file.

Change Log
All notable changes to the "VC65X" extension will be documented in this file.

#{Release.ReleaseName}#
Root node has folder icon instead of vsix icon [Issue #5]
Activate extension only on context menu click [Issue #6]
1.0.4
Add icons for PowerShell (ps1, psd1 and psm1 file extensions)
Add icons for YAML (yaml and yml file extensions)
1.0.3
Show notification in VSCode in case of exceptions
Minor improvements
1.0.0
Initial release

src/emulator.js
//
// Emulator MCS6502 System
//

//const path = require('path');
//const fs = require('fs');
//const process = require('process');

//-----------------------------------------------------------------------------------------------//
// Init module
//-----------------------------------------------------------------------------------------------//
// eslint-disable-next-line
BIND(module);

const fs = require('fs');

//-----------------------------------------------------------------------------------------------//
// Required Modules
//-----------------------------------------------------------------------------------------------//
var Constants = require('src/constants');
const createMemory = require('./system/create-memory');

// MOS Technology, Inc.
const MCS6502 = require('./system/cpu/MCS6502_cpu');
// Rockwell Semiconductor Systems, Inc
const R65C02 = require('./system/cpu/R65C02_cpu');
// Western Design Center (WDC)
const W65C02S = require('./system/cpu/W65C02S_cpu');
const W65C816S = require('./system/cpu/W65C816S_cpu');

const MemoryMapper = require('./system/memory-mapper.js');
const createScreenDevice = require('./system/peripherals/screen-device');
const createUartDevice = require('./system/peripherals/uart-device');
const createVDUDevice = require('./system/peripherals/tms9918');
const createPIADevice = require('./system/peripherals/pia-device');

function getTime() {
    var t = process.hrtime();
    return t[0] * 1000000 + ((t[1] / 1000) | 0);
}

//-----------------------------------------------------------------------------------------------//
// Emulator
//-----------------------------------------------------------------------------------------------//

class Emulator {

    constructor(session) {
        this._session = session;

        this._MM = new MemoryMapper();
        this._memory = createMemory(2 ** 16);
        this._MM.map(this._memory, 0, 0xffff);

        // Map 0xFF bytes of the address space to an "output device" - just stdout
        //this._MM.map(createScreenDevice(), 0x6000, 0x30ff, true);
        this._MM.map(createPIADevice(this._session), 0xC000, 0xC100, true);
        this._MM.map(createVDUDevice(this._session), 0xC200, 0xC201, true);

        this._writableBytes = new Uint8Array(this._memory.buffer);


        //this._cpu = new CPU65C02(this._MM);

        //this.init();
    }

    init(forcedCpuArchitecture) {

        if (null == this._cpu) {
            if (null != forcedCpuArchitecture) {
                switch (forcedCpuArchitecture) {
                    case "6502":
                    case "MCS6502":
                        this._cpu = new MCS6502(this._MM);
                        break;

                    case "R65C02":
                        this._cpu = new R65C02(this._MM);
                        break;

                    case "W65C02S":
                        this._cpu = new W65C02S(this._MM);
                        break;

                    case "W65C816S":
                        this._cpu = new W65C816S(this._MM);
                        break;

                    default:
                        break;
                }

            } else {
                this._cpu = new MCS6502(this._MM);
            }
        }
        this._running = false;
        this._prg = null;
    }

    on(eventName, eventFunction) {
        if (null == this._eventMap) {
            this._eventMap = [];
        }

        this._eventMap[eventName] = eventFunction;
    }

    fireEvent(eventName, arg1, arg2, arg3) {
        if (null == this._eventMap) return null;

        var eventFunction = this._eventMap[eventName];
        if (null == eventFunction) return null;

        return eventFunction(arg1, arg2, arg3);
    }

    getStats() {

        var stats = {
            PC: this._cpu.PC,
            registers: {
                A: this._cpu.A,
                X: this._cpu.X,
                Y: this._cpu.Y,
                S: this._cpu.S
            },
            flags: {
                N: this._cpu.N,
                Z: this._cpu.Z,
                B: this._cpu.B,
                C: this._cpu.C,
                V: this._cpu.V,
                I: this._cpu.I,
                D: this._cpu.D
            },
            irq: this._cpu.irq,
            nmi: this._cpu.nmi,
            opcode: this._cpu.opcode,
            cycles: this._cpu.cycles,
        };

        return stats;
    }

    async start(continueExecution) { // jshint ignore:line

        this._running = true;
        var thisInstance = this;

        var promise = new Promise(function (resolve, reject) {
            setTimeout(function () {
                thisInstance.run(true, resolve, continueExecution);
            }, 0);
        });

        return promise;
    }

    stop() { // jshint ignore:line

        this._running = false;

    }

    run(runAsync, resolve, continueExecution) {

        var result = {
            reason: Constants.InterruptReason.UNKNOWN
        };

        var session = this._session;

        var breakpoints = session._breakpoints;
        var breakpointIndex = 0;
        var nextBreakpoint = -1;

        var statementCounter = 0;

        var firstStepWithoutBreakpoint = (continueExecution ? true : false);

        var lastPC = 0;

        // execution is interrupted after a defined amount of time
        // to let JS proceed with other tasks from the queue
        var startTime = getTime();
        var endTime = 0;
        if (Constants.EmulatorIterationExecutionTime > 0) {
            endTime = startTime + Constants.EmulatorIterationExecutionTime * 1000;
        }
        var checkCounter = 0;

        while (true == this._running) {

            var pc = this._cpu.PC;

            if (null != breakpoints && !firstStepWithoutBreakpoint) {
                if (pc < lastPC) {
                    breakpointIndex = 0;
                }

                while (breakpointIndex < breakpoints.length &&
                    breakpoints[breakpointIndex].address.address < pc) {
                    breakpointIndex++;
                }

                var breakpoint = breakpoints[breakpointIndex];
                if (null != breakpoint && pc == breakpoint.address.address) {
                    if (null != breakpoint.logMessage) {
                        this.fireEvent('logpoint', breakpoint);
                    } else {
                        this.fireEvent('breakpoint', breakpoint);
                        result.reason = Constants.InterruptReason.BREAKPOINT;
                        break;
                    }
                }
            }

            lastPC = pc;

            //this.log();
            this._cpu.step();

            if (this.B) {
                this.fireEvent('break', pc);
                result.reason = Constants.InterruptReason.BREAK;
                break;
            }

            if (true == this.returnReached) {
                result.reason = Constants.InterruptReason.EXIT;
                break;
            }

            statementCounter++;
            if (Constants.EmulatorIterationMaxSteps > 0 &&
                statementCounter > Constants.EmulatorIterationMaxSteps) {
                result.reason = Constants.InterruptReason.YIELD;
                break;
            }

            checkCounter++;
            if (checkCounter >= 1000) {
                checkCounter = 0;
                if (endTime > 0) {
                    var currentTime = getTime();
                    if (currentTime >= endTime) {
                        //console.log("STATEMENTS BEFORE YIELD: " + statementCounter);
                        result.reason = Constants.InterruptReason.YIELD;
                        break;
                    }
                }
            }

            firstStepWithoutBreakpoint = false;
        }

        if (!this._running) {
            result.reason = Constants.InterruptReason.INTERRUPTED;
        }

        if (this._running && result.reason == Constants.InterruptReason.YIELD) {
            var thisInstance = this;
            if (Constants.EmulatorIterationSleepTime > 0) {
                setTimeout(function () { thisInstance.run(true, resolve); }, Constants.EmulatorIterationSleepTime);
            } else {
                process.nextTick(function () { thisInstance.run(true, resolve); });
            }
        } else if (resolve) {
            resolve(result);
        } else {
            return result;
        }
    }

    reset(startAddress) {

        this._cpu.reset();

        this._writableBytes.fill(0);

        if (null != startAddress) {
            // set reset vector to start address
            this.write(0xFFFD, (startAddress >> 8) & 0xFF);
            this.write(0xFFFC, (startAddress & 0xFF));
        }

        this._cpu.S = 0xFF; // initialize stack pointer
        this._cpu.PC = startAddress;
        this._cpu.opcode = this.read(this._cpu.PC);
        this._cpu.cycles = 0;
    }

    /**
     * Execute a single opcode
     */
    step() {
        this._cpu.step();
    }


    read(addr) {

        if (addr < 0 || addr > 0xFFFF) {
            throw new Error('Illegal memory read at address: ' + addr.toString(16).toLowerCase());
        }

        //return this._memory[addr] & 0xFF;
        return this._writableBytes[addr];
    }

    write(addr, value) {
        if (addr < 0 || addr > 0xFFFF) {
            throw new Error('Illegal memory read at address: ' + addr.toString(16).toLowerCase());
        }
        //this._memory[addr] = (value & 0xFF);
        this._writableBytes[addr] = value;
    }

    loadProgram(filename, autoOffsetCorrection, forcedLoadAddress, forcedStartAddress) {

        var prg = null;

        try {
            prg = fs.readFileSync(filename);
        } catch (err) {
            if (err.code == 'ENOENT') {
                throw ("file " + filename + " does not exist");
            } else {
                throw ("unable to read file '" + filename + "'");
            }
        }

        this._prg = prg;

        this.injectProgram(prg, autoOffsetCorrection, forcedLoadAddress, forcedStartAddress);
    }

    injectProgram(prg, autoOffsetCorrection, forcedLoadAddress, forcedStartAddress) {

        var addr;
        var data;
        var startAddr;

        if (null != forcedLoadAddress) {
            addr = forcedLoadAddress;
            data = prg;
        } else {
            addr = ((prg[1] << 8) | prg[0]);
            data = prg.slice(2);
            startAddr = addr;
        }

        if (null != forcedStartAddress) {

            startAddr = forcedStartAddress;

        } else if (true == autoOffsetCorrection) {

            // skip if...
            // starts with valid next statement address
            //        and SYS basic comment
            //        and end of statement zero bytes
            //        !byte $0c,$08,$b5,$07,$9e,$20,$32,$30,$36,$32,$00,$00,$00
            //                              SYS       2   0   6   2

            var addrNextBasic = ((data[1] << 8) | data[0]);
            var maxHeaderBytes = 32;
            var delta = (addrNextBasic - addr);

            if (delta > 0 && delta < maxHeaderBytes) {

                // skip 2 address bytes of next basic token
                var ofs = 2;

                // scanning for "SYS" command 0x9e
                while (ofs < maxHeaderBytes) {
                    if (data[ofs] == 0x9e) break;
                    ofs++;
                }

                ofs++;

                // skip optional spaces
                while (data[ofs] == 0x20) {
                    ofs++;
                }

                // parse call address
                var addressCalc = 0;
                while (ofs < maxHeaderBytes) {
                    var c = data[ofs];
                    if (c < 0x30 || c > 0x39) break;
                    addressCalc = addressCalc * 10 + (data[ofs] - 0x30);
                    ofs++;
                }

                startAddr = addressCalc;
            }
        }

        this.reset(startAddr || 0);
        this._writableBytes.set(data, addr);
        this._cpu.opcode = this.read(this._cpu.PC);
    }
}

//-----------------------------------------------------------------------------------------------//
// Module Exports
//-----------------------------------------------------------------------------------------------//

module.exports = Emulator;
